File: leetcode-python-go/792. 匹配子序列的单词数/main.go
package main

import (
	"fmt"
	"sort"
)

func numMatchingSubseq(s string, words []string) int {
	pos := [26][]int{}
	for i, c := range s {
		pos[c-'a'] = append(pos[c-'a'], i)
	}
	ans := len(words)
	for _, w := range words {
		if len(w) > len(s) {
			ans--
			continue
		}
		p := -1
		for _, c := range w {
			ps := pos[c-'a']
			j := sort.SearchInts(ps, p+1)
			if j == len(ps) {
				ans--
				break
			}
			p = ps[j]
		}
	}
	return ans
}

func main() {
	s := "abcde"
	words := []string{"a", "bb", "acd", "ace"}
	results := numMatchingSubseq(s, words)
	fmt.Println(results)
}

File: leetcode-python-go/792. 匹配子序列的单词数/main.py
from typing import List
from collections import defaultdict
from bisect import bisect_right

s = "abcde"
words = ["a","bb","acd","ace"]

def numMatchingSubseq(s: str, words: List[str]) -> int:
    pos = defaultdict(list)
    for i, c in enumerate(s):
        pos[c].append(i)
    ans = len(words)
    for w in words:
        if len(w) > len(s):
            ans -= 1
            continue
        p = -1
        for c in w:
            ps = pos[c]
            j = bisect_right(ps, p)
            if j == len(ps):
                ans -= 1
                break
            p = ps[j]
    return ans

print(numMatchingSubseq(s, words))